### PR TITLE
x86: Inline reference system array copy for smaller size

### PR DESCRIPTION
**Pull request by a7ehuo**
This is an enhancement to inline arraycopy for reference arrays if copy size is less than 32 bytes (default), or 64 bytes, or 128 bytes. It leverages the enhancement that has been implemented for int arrays and long arrays in OMR. 

This enhancement can be disabled by setting an env variable `TR_DisableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS` or using option `disableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS`.

The inlinling will take place only if there is no NoArrayStoreCheckArrayCopy and it doesn't require barrier.

- [x] **Depends on https://github.com/eclipse/omr/pull/7400**

### CODE REVIEW

**Comment by a7ehuo** [comment]
This change depends on https://github.com/eclipse/omr/pull/7400. Mark it as WIP until https://github.com/eclipse/omr/pull/7400 is merged and promoted.

@0xdaryl May I ask you to review this change? Thank you!

@vijaysun-omr @hzongaro fyi

**Comment by a7ehuo** [comment]
The two shadowed boxes with red lines are the insert new logic 


Screenshot 2024-07-05 at 10 03 53 AM

**Comment by a7ehuo** [comment]
Remove `WIP` since  https://github.com/eclipse/omr/pull/7400 has been picked up by openj9-omr [openj9 branch](https://github.com/eclipse-openj9/openj9-omr/commits/openj9/).  @0xdaryl Ready for review. Thank you!

**Comment by a7ehuo** [comment]
Here is an example of generated sequence for compressed reference 

```
------------------------------
 n86n     (  0)  treetop                                                                              [0x7f277e915aa0] bci=[-1,6,20] rc=0 vc=406 vn=- li=5 udi=- nc=1
 n85n     (  0)    arraycopy  java/lang/System.arraycopy(Ljava/lang/Object;ILjava/lang/Object;II)V[#432  final native static Method] [flags 0x20500 0x0 ] (noArrayStoreCheckArrayCopy referenceArrayCopy )  [0x7f277e915a50] bci=[-1,6,20] rc=0 vc=406 vn=- li=5 udi=- nc=5 flg=0x8020
 n3n      (  0)      ==>aRegLoad (in &GPR_0016) (X!=0 )
 n5n      (  0)      ==>aRegLoad (in &GPR_0032) (X!=0 X>=0 )
 n77n     (  0)      aladd (in GPR_0128) (internalPtr )                                               [0x7f277e9157d0] bci=[-1,0,20] rc=0 vc=406 vn=- li=5 udi=45600 nc=2 flg=0x8000
 n3n      (  0)        ==>aRegLoad (in &GPR_0016) (X!=0 )
 n28n     (  0)        ==>lsub (in GPR_0064) (highWordZero X>=0 )
 n80n     (  0)      aladd (in GPR_0144) (internalPtr )                                               [0x7f277e9158c0] bci=[-1,2,20] rc=0 vc=406 vn=- li=5 udi=46016 nc=2 flg=0x8000
 n5n      (  0)        ==>aRegLoad (in &GPR_0032) (X!=0 X>=0 )
 n32n     (  0)        ==>lsub (in GPR_0080) (highWordZero X>=0 )
 n35n     (  0)      ==>lshl (in GPR_0096) (X>=0 )
------------------------------

 [0x7f277e8fb330]       lea     GPR_0128, qword ptr [&GPR_0016+1*GPR_0064]              # LEA8RegMem
 [0x7f277e8fb4d0]       lea     GPR_0144, qword ptr [&GPR_0032+1*GPR_0080]              # LEA8RegMem
 [0x7f277e8fb960]       mov     GPR_0145, GPR_0128              # MOV8RegReg
 [0x7f277e8fb9f0]       mov     GPR_0146, GPR_0144              # MOV8RegReg
 [0x7f277e8fba80]       mov     GPR_0147, GPR_0096              # MOV8RegReg
 [0x7f277e8fbdb0]       Label L0064:                    # label # (Start of internal control flow)
 [0x7f277e8fbf40]       cmp     dword ptr [GPR_0000+0xa04], 0x00000000  # CMP4MemImms, SymRef [#453 +2564]
 [0x7f277e8fbfd0]       jne     Outlined Label L0066                    # JNE4
 [0x7f277e8fc8d0]       cmp     GPR_0147, 0x00000010    # CMP8RegImm4
 [0x7f277e8fc960]       ja      Label L0070                     # JA4
 [0x7f277e8fca00]       cmp     GPR_0147, 0x00000008    # CMP8RegImm4
 [0x7f277e8fca90]       ja      Label L0069                     # JA4
 [0x7f277e8fcb30]       test    GPR_0147, GPR_0147              # TEST8RegReg
 [0x7f277e8fcbc0]       je      Label L0065                     # JE4
 [0x7f277e8fccf0]       mov     GPR_0148, dword ptr [GPR_0145+1*GPR_0147-0x4]           # L4RegMem, SymRef [#456 -4]
 [0x7f277e8fce10]       mov     GPR_0149, dword ptr [GPR_0145]          # L4RegMem
 [0x7f277e8fcf30]       mov     dword ptr [GPR_0146+1*GPR_0147-0x4], GPR_0148           # S4MemReg, SymRef [#458 -4]
 [0x7f277e8fd050]       mov     dword ptr [GPR_0146], GPR_0149          # S4MemReg
 [0x7f277e8fd0e0]       jmp     Label L0065                     # JMP4
 [0x7f277e8fd180]       Label L0069:                    # label
 [0x7f277e8fd2b0]       mov     GPR_0148, qword ptr [GPR_0145+1*GPR_0147-0x8]           # L8RegMem, SymRef [#460 -8]
 [0x7f277e8fd3d0]       mov     GPR_0149, qword ptr [GPR_0145]          # L8RegMem
 [0x7f277e8fd4f0]       mov     qword ptr [GPR_0146+1*GPR_0147-0x8], GPR_0148           # S8MemReg, SymRef [#462 -8]
 [0x7f277e8fd610]       mov     qword ptr [GPR_0146], GPR_0149          # S8MemReg
 [0x7f277e8fd6a0]       jmp     Label L0065                     # JMP4
 [0x7f277e8fd740]       Label L0070:                    # label
 [0x7f277e8fd7e0]       cmp     GPR_0147, 0x00000020    # CMP8RegImm4
 [0x7f277e8fd870]       ja      Label L0071                     # JA4
 [0x7f277e8fd9a0]       movdqu  VRF_0150, xmmword ptr [GPR_0145+1*GPR_0147-0x10]                # MOVDQURegReg, SymRef [#464 -16]
 [0x7f277e8fdac0]       movdqu  VRF_0151, xmmword ptr [GPR_0145]                # MOVDQURegReg
 [0x7f277e8fdbe0]       movdqu  xmmword ptr [GPR_0146+1*GPR_0147-0x10], VRF_0150                # MOVDQUMemReg, SymRef [#466 -16]
 [0x7f277e8fdd00]       movdqu  xmmword ptr [GPR_0146], VRF_0151                # MOVDQUMemReg
 [0x7f277e8fdd90]       jmp     Label L0065                     # JMP4
 [0x7f277e8fde30]       Label L0071:                    # label
 [0x7f277e8fded0]       cmp     GPR_0147, 0x00000040    # CMP8RegImm4
 [0x7f277e8fdf60]       ja      Label L0068                     # JA4
 [0x7f277e8fe090]       vmovdqu VRF_0150, ymmword ptr [GPR_0145+1*GPR_0147-0x20]                # VMOVDQUYmmMem, SymRef [#468 -32]
 [0x7f277e8fe1b0]       vmovdqu VRF_0151, ymmword ptr [GPR_0145]                # VMOVDQUYmmMem
 [0x7f277e8fe2d0]       vmovdqu ymmword ptr [GPR_0146+1*GPR_0147-0x20], VRF_0150                # VMOVDQUMemYmm, SymRef [#470 -32]
 [0x7f277e8fe3f0]       vmovdqu ymmword ptr [GPR_0146], VRF_0151                # VMOVDQUMemYmm
 [0x7f277e8fe480]       jmp     Label L0065                     # JMP4
 [0x7f277e8fe520]       Label L0068:                    # label
 [0x7f277e8fe620]       sub     GPR_0146, GPR_0145              # SUB8RegReg
 [0x7f277e8fe6b0]       cmp     GPR_0146, GPR_0147              # CMP8RegReg
 [0x7f277e8fe7d0]       lea     GPR_0146, qword ptr [GPR_0146+1*GPR_0145]               # LEA8RegMem
 [0x7f277e8fe860]       jb      Outlined Label L0073                    # JB4
 [0x7f277e8ff000]       shr     GPR_0147, 0x02  # SHR8RegImm1
 [0x7f277e8ff090]       rep movsd                       # REPMOVSD
 [0x7f277e8ff370]       assocreg                        # assocreg

```

**Comment by a7ehuo** [comment]
As per discussion with @vijaysun-omr, this feature is enabled by default in this PR and can be disabled by setting an env variable `TR_DisableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS` or using option `disableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS`

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -1887,11 +1887,42 @@ TR::Register *J9::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Cod
...
+         tmpXmmYmmReg2 = cg->allocateRegister(TR_VRF);
+         }
+
       generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, RSI, srcReg, cg);
       generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, RDI, dstReg, cg);
       generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, RCX, sizeReg, cg);
 
-      auto deps = generateRegisterDependencyConditions((uint8_t)5, 5, cg);
+      TR::RegisterDependencyConditions* deps = enableInlineForSmallSize ? generateRegisterDependencyConditions((uint8_t)9, 9, cg) :
+                                                                          generateRegisterDependencyConditions((uint8_t)5, 5, cg);
```
Comment (0xdaryl) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:1924]: You can keep the generated code smaller by:
```
   int8_t numDeps = enableInlineForSmallSize ? 9 : 5;
   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions(numDeps, numDeps, cg);
```

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -1887,11 +1887,42 @@ TR::Register *J9::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Cod
...
 
+      TR::Register* tmpReg1 = NULL;
+      TR::Register* tmpReg2 = NULL;
+      TR::Register* tmpXmmYmmReg1 = NULL;
+      TR::Register* tmpXmmYmmReg2 = NULL;
+
+      static char *enableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS = feGetEnv("TR_EnableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS");
+
+      bool enableInlineForSmallSize = enableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS &&
+                                      !cg->comp()->getOption(TR_DisableReferenceArrayCopyInlineSmallSizeWithoutREPMOVS) &&
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Addressed in [e721562](https://github.com/eclipse-openj9/openj9/compare/21255244442efda857a4a2c7772e1c6204a8d084..e7215625c5684fa2a4397eb5d65d1bf3a3eb8c60)

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (a7ehuo) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Addressed in [e721562](https://github.com/eclipse-openj9/openj9/compare/21255244442efda857a4a2c7772e1c6204a8d084..e7215625c5684fa2a4397eb5d65d1bf3a3eb8c60)

**Comment by a7ehuo** [comment]
@0xdaryl All comments are addressed. Ready for another review. Thank you!

**Comment by 0xdaryl** [comment]
Jenkins test sanity xlinux,win,xmac jdk21

**Review by 0xdaryl — APPROVED** [review]
Windows build nodes are offline.  I am satisfied that the Linux and macOS testing would have caught any issues.

**Pull request merged**